8081474: SwingWorker calls 'done' before the 'doInBackground' is finished

prsadhuk opened the pull request:
SwingWorker done() method [spec ](https://github.com/openjdk/jdk/blob/master/src/java.desktop/share/classes/javax/swing/SwingWorker.java#L452) says "Executed on the Event Dispatch Thread after the doInBackground method is finished"
but there's no mechanism in place to honor that claim.
The [spec](https://github.com/openjdk/jdk/blob/master/src/java.desktop/share/classes/javax/swing/SwingWorker.java#L289)
also says the state should be DONE after doInBackground() returns which is also not done.

Modified the code to honour the specification.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issues
 * [JDK-8081474](https://bugs.openjdk.org/browse/JDK-8081474): SwingWorker calls 'done' before the 'doInBackground' is finished
 * [JDK-8301940](https://bugs.openjdk.org/browse/JDK-8301940): SwingWorker calls 'done' before the 'doInBackground' is finished (**CSR**) (Withdrawn)


### Reviewers
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**)
 * [Sergey Bylokhov](https://openjdk.org/census#serb) (@mrserb - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11940/head:pull/11940` \
`$ git checkout pull/11940`

Update a local copy of the PR: \
`$ git checkout pull/11940` \
`$ git pull https://git.openjdk.org/jdk pull/11940/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11940`

View PR using the GUI difftool: \
`$ git pr show -t 11940`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11940.diff">https://git.openjdk.org/jdk/pull/11940.diff</a>

</details>

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/SwingWorker.java:
```diff
@@ -751,8 +751,22 @@ public void run() {
+            if (state != StateValue.DONE) {
+                do {
+                    try {
+                        Thread.sleep(100);
```
I agree it doesn't look good. If adding `sleep` is needed, there's something wrong.

prsadhuk commented:
Any more comments on this? 
Should we close the JBS as "Not an issue"?
or
Should we just update the spec citing it's not necessary to wait for doInBackground() to finish before done() is called?
or
any other suggestion for the fix?

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/SwingWorker.java:
```diff
@@ -301,18 +301,18 @@ public SwingWorker() {
                         setState(StateValue.STARTED);
-                        return doInBackground();
+                        T ret = doInBackground();
+                        setState(StateValue.DONE);
```
What if doInBackground thrown an exception? `DONE` state will never be reached.

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/SwingWorker.java:
```diff
@@ -753,6 +753,13 @@ public void run() {
         if (SwingUtilities.isEventDispatchThread()) {
             doDone.run();
         } else {
+            if (state != StateValue.DONE) {
+                do {
+                    try {
+                        Thread.sleep(100);
+                    } catch (InterruptedException e) {}
+                } while (state != StateValue.DONE);
+            }
```
In the previous iteration, using `sleep` for waiting was *the concern*, you're still using `sleep`.

This is not going to work because it makes `cancel` wait until `DONE` state is reached, which is not what one would expect, especially taking into account that `cancel` is likely called from EDT to cancel the background operation and blocking EDT is not acceptable.

prsadhuk replied:
I removed sleep from EDT case and not blocking EDT and guess sleep is now being done for non-EDT case...
If it is to be called from EDT then it should go to "if" block and not to "else" which is what I based my fix on..

Anyway, I appreciate your fix and will see to it..

aivanov-jdk replied:
Yet having `sleep` is an indication of busy wait. You could've used `CountDownLatch` in conjunction with `setState` to avoid `sleep` at all.

> I removed sleep from EDT case and not blocking EDT and guess sleep is now being done for non-EDT case...

Ah, right. But then if `cancel` is called on EDT, `done` is invoked before `doInBackground` completes, isn't it?

If `cancel` is called from another thread, the main thread as in the test, it doesn't return until `doInBackground` completes. With your test case and fix, `cancel` blocks for 5 seconds.

Either way, the behaviour does not look right.

prsadhuk replied:
It seems the isDone method modification is causing a JCK test to fail and there's no spec for it to challenge JCK test, although logically it seems right that STATE should be DONE but for now, I have modified the fix which satisfies regression test and JCK both..

prsadhuk replied:
isDone() spec mention `"Completion may be due to normal termination, an exception, or cancellation -- in all of these cases, this method will return true"`
so for cancel state also, the state needs to be in DONE state, in which case we can reinstate the DONE state check in isDone()

aivanov-jdk replied:
`isDone()` spec is inherited from `Future.isDone()` and it is true for how `FutureTask` is implemented.

https://github.com/openjdk/jdk/blob/810c8a271b4524ae776e2306ef699e04a7d145a2/src/java.desktop/share/classes/javax/swing/SwingWorker.java#L565-L570

Can't we change it?

Given what is said in the spec for `SwingWorker` before, I assume `isDone` should return `true` when the status of `SwingWorker` moves to `DONE` and it should occur when `doInBackground` returns as this is defined in the spec:

https://github.com/openjdk/jdk/blob/810c8a271b4524ae776e2306ef699e04a7d145a2/src/java.desktop/share/classes/javax/swing/SwingWorker.java#L288-L293

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestSwingWorker.java:
```diff
@@ -0,0 +1,77 @@
+ */
+import javax.swing.SwingWorker;
+
+public class TestSwingWorker {
```
I believe it was decided to give test classes a meaningful names but `TestSwingWorker` isn't better than `bug8081474`. The latter has an advantage though, it's unique.

`DoneBeforeDoInBackground` seems a better name which describes the test purpose, don't you think?

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestSwingWorker.java:
```diff
@@ -0,0 +1,77 @@
+                        System.out.println("Working...");
+                        Thread.sleep(1000);
+                    }
+                }
+                catch (InterruptedException ex) {
```
```suggestion
                } catch (InterruptedException ex) {
```
`catch` should be on the same line as the closing `brace` of the previous block.

aivanov-jdk reviewed (changes requested):
Blocking EDT with `cancel` is unacceptable.

I propose the following fix ([**diff to your PR**](https://github.com/openjdk/jdk/compare/pr/11940...aivanov-jdk:prasanta/8081474-SwingWorker-done)):
```diff
diff --git a/src/java.desktop/share/classes/javax/swing/SwingWorker.java b/src/java.desktop/share/classes/javax/swing/SwingWorker.java
index 6acca76d859..9da72fe2f0a 100644
--- a/src/java.desktop/share/classes/javax/swing/SwingWorker.java
+++ b/src/java.desktop/share/classes/javax/swing/SwingWorker.java
@@ -301,18 +301,16 @@ public abstract class SwingWorker<T, V> implements RunnableFuture<T> {
                 new Callable<T>() {
                     public T call() throws Exception {
                         setState(StateValue.STARTED);
-                        T ret = doInBackground();
-                        setState(StateValue.DONE);
-                        return ret;
+                        try {
+                            return doInBackground();
+                        } finally {
+                            setState(StateValue.DONE);
+                            doneEDT();
+                        }
                     }
                 };
 
-        future = new FutureTask<T>(callable) {
-                       @Override
-                       protected void done() {
-                           doneEDT();
-                       }
-                   };
+       future = new FutureTask<T>(callable);
        state = StateValue.PENDING;
        propertyChangeSupport = new SwingWorkerPropertyChangeSupport(this);
        doProcess = null;
@@ -566,7 +564,7 @@ public abstract class SwingWorker<T, V> implements RunnableFuture<T> {
      * {@inheritDoc}
      */
     public final boolean isDone() {
-        return future.isDone();
+        return future.isDone() && state == StateValue.DONE;
     }
 
     /**
@@ -753,13 +751,6 @@ public abstract class SwingWorker<T, V> implements RunnableFuture<T> {
         if (SwingUtilities.isEventDispatchThread()) {
             doDone.run();
         } else {
-            if (state != StateValue.DONE) {
-                do {
-                    try {
-                        Thread.sleep(100);
-                    } catch (InterruptedException e) {}
-                } while (state != StateValue.DONE);
-            }
             doSubmit.add(doDone);
         }
     }
```

or diff to master:
```diff
diff --git a/src/java.desktop/share/classes/javax/swing/SwingWorker.java b/src/java.desktop/share/classes/javax/swing/SwingWorker.java
index 0d86075be3f..9da72fe2f0a 100644
--- a/src/java.desktop/share/classes/javax/swing/SwingWorker.java
+++ b/src/java.desktop/share/classes/javax/swing/SwingWorker.java
@@ -301,18 +301,16 @@ public abstract class SwingWorker<T, V> implements RunnableFuture<T> {
                 new Callable<T>() {
                     public T call() throws Exception {
                         setState(StateValue.STARTED);
-                        return doInBackground();
+                        try {
+                            return doInBackground();
+                        } finally {
+                            setState(StateValue.DONE);
+                            doneEDT();
+                        }
                     }
                 };
 
-        future = new FutureTask<T>(callable) {
-                       @Override
-                       protected void done() {
-                           doneEDT();
-                           setState(StateValue.DONE);
-                       }
-                   };
-
+       future = new FutureTask<T>(callable);
        state = StateValue.PENDING;
        propertyChangeSupport = new SwingWorkerPropertyChangeSupport(this);
        doProcess = null;
@@ -566,7 +564,7 @@ public abstract class SwingWorker<T, V> implements RunnableFuture<T> {
      * {@inheritDoc}
      */
     public final boolean isDone() {
-        return future.isDone();
+        return future.isDone() && state == StateValue.DONE;
     }
 
     /**
```

This approach ensures `cancel` completes quickly and `done` is called only after `doInBackground` exits.

I also suggest adding a condition to the test which verifies `cancel` doesn't take too long as well as ensuring `done` was called before exiting from `main`.

The `isDone` method has the same problem: it returns `true` before the background task has completed, that's why I modified it so that it returns `true` only when the state is `DONE` too.

I believe the time of `sleep` to simulate work can be reduced for automatic testing, however, I agree a few seconds' delay is better for debugging. That's why the time could be configurable in constants.

aivanov-jdk commented:
> Should we just update the spec citing it's not necessary to wait for `doInBackground()` to finish before `done()` is called?

It is also an option.

All the suggested approaches set the state to `DONE` before `done` is called now. Yet the Workflow states the following: “After the `doInBackground` method is finished the `done` method is executed. Then `SwingWorker` notifies any `PropertyChangeListeners` about the state property change to `StateValue.DONE`.”

However, the last sentence may not have been guaranteed before either but it was likely satisfied as `doneEDT` was called before `setState(StateValue.DONE)`. So, `done` had likely been executed before `PropertyChangeListener` was notified.

Hence, a small update to the spec may still be required whatever option we choose.

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/SwingWorker.java:
```diff
@@ -552,6 +550,7 @@ public final void execute() {
      * {@inheritDoc}
      */
     public final boolean cancel(boolean mayInterruptIfRunning) {
+        setState(StateValue.DONE);
         return future.cancel(mayInterruptIfRunning);
```
Setting the state to `DONE` unconditionally doesn't seem right either. What if `doInBackground` isn't started? What if `doInBackground` starts after you compare the state is `PENDING`?

If `cancel` changes the state, it must do it after `future.cancel` returns. But then, the state moves to `DONE` before `doInBackground` exited which contradicts the spec for `DONE`.

prsadhuk replied:
OK..I have removed unconditional setting of DONE state..
But then as I mentioned, it would cause JCK failure, so I have reverted isDone() change too...

aivanov-jdk replied:
> OK..I have removed unconditional setting of DONE state.. But then as I mentioned, it would cause JCK failure, so I have reverted isDone() change too...

Should we update the spec for `isDone()`? I think it's the right way because it makes the behaviour consistent. Doing so requires a CSR which will be the grounds for updating the JCK test.

prsadhuk replied:
I am not sure if it is entirely wrong..I will keep it as it is for now..I dont want to go in CSR route for now..
The present iteration takes care of reproducer and also satisifies JCK..
If in future it is needed, we can take a look..

aivanov-jdk replied:
Let's see what others have to say.

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/SwingWorker.java:
```diff
@@ -566,7 +565,7 @@ public final boolean isCancelled() {
      */
     public final boolean isDone() {
-        return future.isDone();
+        return future.isDone() && state == StateValue.DONE;
```
```suggestion
        return state == StateValue.DONE;
```
I guess we can remove the call to `future.isDone()` altogether. Either both conditions are true, or both are false.

prsadhuk replied:
It will cause JCK issue..

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -0,0 +1,75 @@
+                }
+                try {
+                    System.out.println("Cleaning up");
+                    Thread.sleep(5000);
```
I still think that you should define constants for *all the delays*, and using 1 or 2 seconds for cleaning should be enough.

prsadhuk replied:
done

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -0,0 +1,75 @@
+        worker.execute();
+        Thread.sleep(5000);
+
+        worker.cancel(true);
```
I still suggest ensuring `cancel` does not take too long.

prsadhuk replied:
Not sure I understand this...please elaborate...

aivanov-jdk replied:
Something like this:

https://github.com/aivanov-jdk/jdk/blob/15580d0f1a3a9630b224bb567aacec03d825c777/test/jdk/javax/swing/SwingWorker/TestSwingWorker.java#L83-L90

You can take all the changes to the test as it is quicker and configurable.

prsadhuk replied:
Thanks for the test. I have updated..
Also, as per the spec you mentioned
https://github.com/openjdk/jdk/blob/810c8a271b4524ae776e2306ef699e04a7d145a2/src/java.desktop/share/classes/javax/swing/SwingWorker.java#L288-L293

I have added one more condition

turbanoff commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -0,0 +1,75 @@
+
+public class TestDoneBeforeDoInBackground {
+
+    static boolean doInBackground =  false;
```
```suggestion
    static boolean doInBackground = false;
```

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -0,0 +1,90 @@
+
+public class TestDoneBeforeDoInBackground {
+
+    static boolean doInBackground = false;
+    static final int WAIT_TIME = 200;
```
Why not `private`?

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -0,0 +1,90 @@
+            throw new RuntimeException("Cancel took too long: "
+                                       + ((end - start) / 1000.0d) + " s");
+        }
+        if (doInBackground && worker.getState() != SwingWorker.StateValue.DONE) {
+            throw new RuntimeException("doInBackground is finished " +
+                                       " but State is not DONE");
+        }
```
I think you should wait for cleanup to finish and ensure that `doInBackground` is set to `true` and the state is `DONE`.

If you don't wait until cleanup wait completes, the test does not test anything because jtreg will shutdown the JVM as soon main thread exists, which will happen before `done` gets called.

prsadhuk replied:
Dont think so...If `doInBackground `is true it should mean cleanup is done as `doInBackground ` is false by default

aivanov-jdk replied:
That's the thing. If `cancel` doesn't wait until `doInBackground` completes, which *it shouldn't do*, then `cancel` returns quickly. Therefore the main thread reaches this condition before the “cleanup” in `doInBackground` finishes, as such the test should always fail.

Do I miss anything?

prsadhuk replied:
ok..understood...modified...

mrserb commented:
How the updated code is supposed to work if the worker will have some state updated by the "doInBackground" and used by the "done" method. I guess if the "doInBackground" will be canceled the "done" method may throw an exception? Do we have such usage in our codebase?

prsadhuk commented:
The scenario is covered by JCK test and present PR iteration does not show any issue

mrserb commented:
> SwingWorker is used in JEditorPane as well as in jconsole packages. Yet I didn't look into the details how it's used.

I think all of that usages by jconsole will start to throw the "CancellationException" in the "done" method since now we will start to call it. Not sure can it cause some issue or not.

aivanov-jdk commented:
> > SwingWorker is used in JEditorPane as well as in jconsole packages. Yet I didn't look into the details how it's used.
> 
> I think all of that usages by jconsole will start to throw the "CancellationException" in the "done" method since now we will start to call it. Not sure can it cause some issue or not.

Why do you think so?

The `done` method hasn't thrown `CancellationException`, and it will not. If it had thrown the exception, Prasanta's test would've caught it.

Additionally, the `done` method *must* be called. It was called before, it is called now.

I cannot understand your concern.

prsadhuk commented:
Implementation of `done` method is not changed to cause/invoke `CancellationException `.
 As per [JDK-6608234](https://bugs.openjdk.org/browse/JDK-6608234) "CancellationException" will be called for `get` method if `cancel `is invoked, which is still valid/working and as mentioned, we have not seen any exception either in jtreg or in jck with updated fix.

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -0,0 +1,96 @@
+
+    static boolean doInBackground = false;
+    private static final int WAIT_TIME = 200;
+    private static final long CLEANUP_TIME = 3000;
```
```suggestion
    private static final long CLEANUP_TIME = 1000;
```
1 second is enough for automatic test.

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -0,0 +1,96 @@
+
+public class TestDoneBeforeDoInBackground {
+
+    static boolean doInBackground = false;
```
```suggestion
    private static boolean doInBackground = false;
```
Let's make it private too.

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -81,6 +83,10 @@ protected void done() {
+        if (!doneLatch.await(CLEANUP_TIME + 1000L, TimeUnit.MILLISECONDS)) {
+            throw new RuntimeException("done didn't complete in time");
+        }
+        System.out.println("doInBackground " + doInBackground);
```
```suggestion
        System.out.println("doInBackground = " + doInBackground);
```
Maybe add getState() too?

aivanov-jdk reviewed (approved):
Looks good to me.

mrserb commented:
> Why do you think so?
> 
> The `done` method hasn't thrown `CancellationException`, and it will not. If it had thrown the exception, Prasanta's test would've caught it.

But the `done` methods which call "get" will start to throw that exception, right? Implementation of "done" method in the jconsole actually calls "get". Is that pattern could be common?

mrserb commented on src/java.desktop/share/classes/javax/swing/SwingWorker.java:
```diff
@@ -301,18 +301,16 @@ public SwingWorker() {
+                        } finally {
+                            setState(StateValue.DONE);
+                            doneEDT();
+                        }
```
Probably been discussed already, but this will not follow this part of spec:
```
 * Before the {@code doInBackground} method is invoked on a <i>worker</i> thread,
 * {@code SwingWorker} notifies any {@code PropertyChangeListeners} about the
 * {@code state} property change to {@code StateValue.STARTED}.  After the
 * {@code doInBackground} method is finished the {@code done} method is
 * executed.  Then {@code SwingWorker} notifies any {@code PropertyChangeListeners}
 * about the {@code state} property change to {@code StateValue.DONE}.
```

aivanov-jdk replied:
Valid concern.

I mentioned it before [here](https://github.com/openjdk/jdk/pull/11940#issuecomment-1407690445), yet I forgot about this later on.

Changing the order of the calls should be enough.

I guess we should verify this condition in the test, what do you think?

mrserb replied:
But it did not work as expected before the fix as well since the "notification and done" work in parallel anyway.

mrserb replied:
It could be tested by flushing and blocking the EDT in the listener. That will allow us to check the order of events posted from the listener and from the "doneEDT".

aivanov-jdk replied:
I haven't verified it, but I think it worked as it's stated: `done` is called first, then listeners are notified. According to the order of calls

https://github.com/openjdk/jdk/blob/b00b70c2400d28070d26630614a010bc52237827/src/java.desktop/share/classes/javax/swing/SwingWorker.java#L310-L313

`SwingWorker.done` gets invoked first followed by `setState` which, in its turn, notifies the listeners.

At this time, as you pointed out, the order is reversed:

https://github.com/openjdk/jdk/blob/76b02ea3a8add954705d6df7f91f5896d141b02d/src/java.desktop/share/classes/javax/swing/SwingWorker.java#L307-L308

aivanov-jdk replied:
> It could be tested by flushing and blocking the EDT in the listener. That will allow us to check the order of events posted from the listener and from the "doneEDT".

Is flushing required?

Two `AtomicBoolean`s should do the job. EDT can't process two events simultaneously, one follows the other. So, `done` is expected to be called while both are false, and it sets its own flag to `true`; the listener should be called second, so its flag is false and `done`'s flag is true.

I'll leave it to Pransanta to figure out the details.

prsadhuk replied:
It seems the order of sequence is
listener->State.STARTED->doInBackground->listener->DONE->done

aivanov-jdk replied:
> It seems the order of sequence is listener->State.STARTED->doInBackground->listener->DONE->done

Yes, but the different order is specified: listener(STARTED) -> doInBackground -> done -> listener(DONE).

prsadhuk replied:
But then it will violate
https://github.com/openjdk/jdk/blob/810c8a271b4524ae776e2306ef699e04a7d145a2/src/java.desktop/share/classes/javax/swing/SwingWorker.java#L288-L293

aivanov-jdk replied:
Hm… not really, state will transition to `DONE` after `doInBackground` is finished.

In fact, it'll happen soon enough: `doneEDT` will schedule execution of `done` on EDT and exit. Then the state will change to `DONE`. Yet listeners are notified on EDT, to this notification will also be scheduled to run on EDT.

Very much likely that the state is `DONE` when `done` is executed but it's not guaranteed.

---

The specification isn't clear, and some parts somewhat contradict each other. Logically, the state should transition to `DONE` as soon as `doInBackground` completes. Then both `done` method and listeners are called when the state is `DONE` that is *the background work is complete*.

Yet as it's specified now, the `done` method is called while the state is still `STARTED`. This kind of implies calling `get` may block because the result of computation may not be available yet. However, it's not the case: `get` never blocks if called from `done`: it either returns the result or throws `CancellationException`.

mrserb replied:
> The specification isn't clear, and some parts somewhat contradict each other. Logically, the state should transition to DONE as soon as doInBackground completes. Then both done method and listeners are called when the state is DONE that is the background work is complete.

Which parts are contradicts? I mean that the old implementation call DONE-listener last, as specified in the header of the class what part of the spec mention different behavior? Above we discussed that it is possible to see a difference if the listener will do some work on EDT, after the fix the listener will not be called last, is that safe to assume it does not break something?

aivanov-jdk replied:
> > The specification isn't clear, and some parts somewhat contradict each other. Logically, the state should transition to `DONE` as soon as `doInBackground` completes. Then both done method and listeners are called when the state is `DONE` that is the background work is complete.
> 
> Which parts are contradicts?

This was a wrong statement. In fact, I referred to the problem that is being fixed here: `done` may be called before `doInBackground`.

https://github.com/openjdk/jdk/blob/0aeebee284effe9abd0ed3cf2845430b40bb53bd/src/java.desktop/share/classes/javax/swing/SwingWorker.java#L288-L293

The spec for `DONE` state does not specify when the transition occurs.

> I mean that the old implementation call DONE-listener last, as specified in the header of the class what part of the spec mention different behavior?

Yes, this is how it's specified now.

> Above we discussed that it is possible to see a difference if the listener will do some work on EDT, after the fix the listener will not be called last, is that safe to assume it does not break something?

What kind of difference? Yes, you can observe the fact that now the `DONE`-state-listener is called before the `done` method.

In either case, the listener and the method are called when `get` does not block. The background job is finished when either of these events occurs.

I see no problem here.

aivanov-jdk replied:
> > It could be tested by flushing and blocking the EDT in the listener. That will allow us to check the order of events posted from the listener and from the "doneEDT".
> 
> Is flushing required?
> 
> Two `AtomicBoolean`s should do the job.

I referred to `isStateDoneNotified` and `isDoneCalled`.

> EDT can't process two events simultaneously, one follows the other. So, `done` is expected to be called while both are `false`, and it sets its own flag to `true`; the listener should be called second, so its flag is `false` and `done`'s flag is `true`.

Do we want to ensure it?

mrserb commented:
> > Why do you think so?
> > The `done` method hasn't thrown `CancellationException`, and it will not. If it had thrown the exception, Prasanta's test would've caught it.
> 
> But the `done` methods which call "get" will start to throw that exception, right? Implementation of "done" method in the jconsole actually calls "get". Is that pattern could be common?

I see your point now, yes that should not be a problem.

aivanov-jdk commented:
> > Why do you think so?
> > The `done` method hasn't thrown `CancellationException`, and it will not. If it had thrown the exception, Prasanta's test would've caught it.
> 
> But the `done` methods which call "get" will start to throw that exception, right? Implementation of "done" method in the jconsole actually calls "get". Is that pattern could be common?

`done` which calls `get` **will *continue*** to see this exception: nothing has changed. As soon as `cancel` is called, `get` throws `CancellationException` (unless `doInBackground` completed by that time). The old code threw the exception just like the new one does.

You can take Prasanta's test and add a call to `get` into `done` — it throws `CancellationException` if run with jdk19 or 21 without the fix. It still throws `CancellationException` with the fix. (I've just done it to confirm my own sanity. You can do it too.)

Nothing has changed in this regard.

aivanov-jdk commented:
/reviewers 2

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -0,0 +1,97 @@
+    private static final CountDownLatch doneLatch = new CountDownLatch(1);
+
+    public static void main(String[] args) throws InterruptedException {
+        SwingWorker<String, String> worker =
+            new SwingWorker<String, String>() {
```
```suggestion
        SwingWorker<String, String> worker = new SwingWorker<>() {
```
Just noticed, it can be simplified.

aivanov-jdk commented:
The more I think about it, the more I'm inclined towards this change may be unnecessary at all. Yet the spec needs to be updated.

Consider the following scenario. The user starts a long running operation, so `doInBackground` is being executed. It may block access to UI and show the progress. Then they decide they don't to wait any more, so they cancel the operation by clicking *Cancel* button the UI which calls `cancel(true)` on `SwingWorker`.

The thread that runs `doInBackground` is interrupted. It may not exit yet but *does it matter?*

From the user's point of view, the operation is already cancelled. That fact that the background thread needs to perform some clean-up actions is irrelevant.

From the developer's point of view, it does matter. It depends on what actions are valid on this instance of `SwingWorker`. Until `doInBackground` finishes, a resource may still be busy while it performs its clean-up.

Can the clean-up be run in background too? There's no univocal answer to this question: in some cases, it can; in other cases, it cannot.

Thus, with the current approach in this fix, `SwingWorker` provides stronger guarantees which are stated in the spec that `done` is called only after `doInBackground` finishes whether it's cancelled or not.

Before this fix, the guarantee was that the call to `get` doesn't block when `done` is called or when the listeners are notified about the state `DONE`.

prsadhuk commented:
I think we should update the wording alongside this fix like this

```
  * {@code state} property change to {@code StateValue.STARTED}.  After the
- * {@code doInBackground} method is finished the {@code done} method is
- * executed.  Then {@code SwingWorker} notifies any {@code PropertyChangeListeners}
- * about the {@code state} property change to {@code StateValue.DONE}.
+ * {@code doInBackground} method is finished, then
+ * {@code SwingWorker} notifies any {@code PropertyChangeListeners}
+ *  about the {@code state} property change to {@code StateValue.DONE}, then {@code done} method is
+ *  executed.
```

To me, when `doInBackground `is finished, SwingWorker's job is done and so the STATE should be changed to DONE which is as per spec already mentioned earlier, and then it will call user's `done `method spec of which just says it will be executed `"after the {@code doInBackground} method is finished." ` which will still be valid.

aivanov-jdk commented:
Yeah, I totally agree. It makes more sense to me this way. But it requires a CSR.

> ```
>   * {@code state} property change to {@code StateValue.STARTED}.  After the
> + * {@code doInBackground} method is finished, ~~then~~
> + * {@code SwingWorker} notifies any {@code PropertyChangeListeners}
> + *  about the {@code state} property change to {@code StateValue.DONE}, then {@code done} method is
> + *  executed.
> ```

Remove the first *‘then’*. Maybe re-flow the text. Other than that, it looks good and it's consistent to the spec of `DONE`.

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -72,6 +74,31 @@ protected void done() {
+                               "state STARTED after doInBackground is invoked");
+                    }
+                    // After the doInBackground method is finished
+                    // SwingWorker} notifies PropertyChangeListeners
```
```suggestion
                    // SwingWorker notifies PropertyChangeListeners
```

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -72,6 +74,31 @@ protected void done() {
+                    if (doInBackground &&
+                            worker.getState() != SwingWorker.StateValue.DONE) {
+                        throw new RuntimeException(
+                            "listener called after doInBackground is finised" +
```
```suggestion
                            "PropertyChangeListeners called after doInBackground is finished" +
```

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -27,9 +27,12 @@
 import javax.swing.SwingWorker;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
```
`AtomicBoolean` isn't currently used.

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -0,0 +1,124 @@
+
+public class TestDoneBeforeDoInBackground {
+
+    private static boolean doInBackground = false;
```
```suggestion
    private static volatile boolean doInBackground = false;
```
It still needs to be `volatile`, it's consistently accessed from three different threads.

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -0,0 +1,124 @@
+            @Override
+            protected void done() {
+                if (!doInBackground) {
+                    throw new RuntimeException("done called before doInBackground");
```
```suggestion
                    throw new RuntimeException("done called before doInBackground finished");
```
Otherwise, it looks very confusing.

aivanov-jdk reviewed (changes requested):
The changes to the spec look go to me.

Since a CSR is required, I suggest updating the spec for `isDone` method so that it returns `true` only after `doInBackground` exits. Now `isDone` returns `true` after `cancel` completes but the `DONE` state isn't reached yet.

I think this change will make the implementation more consistent is easier to use.

As for the test, it does not ensure the `STARTED` state is notified before `doInBackground` started running, nor does it ensure `DONE` state is notified before `done` method is called.

Isn't it what we want to test?

If I run the test with jdk19 and comment out the `throw` statement in `done`, the test passes. Yet I expect it to fail since the order of notification about `DONE` state and `done` call is reversed in the fix.

Or do you both think it's too much for the test?

prsadhuk commented:
> Since a CSR is required, I suggest updating the spec for `isDone` method so that it returns `true` only after `doInBackground` exits. Now `isDone` returns `true` after `cancel` completes but the `DONE` state isn't reached yet.

My take is current code fix and spec clarification is done to make code/spec wording consistent with **existing** spec
whereas this change is to **enhance** the code/spec which I think we should keep it separate from this fix, as this will cause JCK test to be excluded and will cause JCK to update the test and I am not sure how that will pan out, so I suggest to keep the change to minimal to the extent only which is needed for this fix..

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -88,14 +88,22 @@ public void propertyChange(PropertyChangeEvent evt) {
-                            "listener called after doInBackground is finised" +
-                            " but before State changed to DONE");
+                            "PropertyChangeListeners called after " +
+                            " doInBackground is finised but before State changed to DONE");
```
```suggestion
                            " doInBackground is finished but before State changed to DONE");
```

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -88,14 +88,22 @@ public void propertyChange(PropertyChangeEvent evt) {
                                "PropertyChangeListeners called with " +
                                "state STARTED after doInBackground is invoked");
                     }
+                    // Ensure the STARTED state is notified
+                    // before doInBackground started running
+                    if (!doInBackground &&
+                         worker.getState() == SwingWorker.StateValue.STARTED) {
```
This condition doesn't guarantee that `STARTED` state is set before `doInBackground` starts running. The boolean flag is set to `true` only in the end of the method.

aivanov-jdk commented:
> > Since a CSR is required, I suggest updating the spec for `isDone` method so that it returns `true` only after `doInBackground` exits. Now `isDone` returns `true` after `cancel` completes but the `DONE` state isn't reached yet.
> 
> My take is current code fix and spec clarification is done to make code/spec wording consistent with **existing** spec whereas this change is to **enhance** the code/spec which I think we should keep it separate from this fix, as this will cause JCK test to be excluded and will cause JCK to update the test and I am not sure how that will pan out, so I suggest to keep the change to minimal to the extent only which is needed for this fix..

Reasonable.

Will you submit a low-priority enhancement then?

aivanov-jdk commented:
I admit I got confused … by now.

It's been a long discussion.

> My take is current code fix and spec clarification is done to make code/spec wording consistent with **existing** spec whereas this change is to **enhance** the code/spec which I think we should keep it separate from this fix, as this will cause JCK test to be excluded and will cause JCK to update the test and I am not sure how that will pan out, so I suggest to keep the change to minimal to the extent only which is needed for this fix..

Yet the spec change is unnecessary if you change the order of calls as [discussed above](https://github.com/openjdk/jdk/pull/11940#discussion_r1095103200).

Referring to [the same thread](https://github.com/openjdk/jdk/pull/11940#discussion_r1095741184),

> > > It seems the order of sequence is listener->State.STARTED->doInBackground->listener->DONE->done
> > 
> > Yes, but the different order is specified: listener(STARTED) -> doInBackground -> done -> listener(DONE).
> >
> But then it will violate
> 
> https://github.com/openjdk/jdk/blob/810c8a271b4524ae776e2306ef699e04a7d145a2/src/java.desktop/share/classes/javax/swing/SwingWorker.java#L288-L293

No, it doesn't violate: the state transitions to `DONE` *after method is finished*.

And I realised [I had addressed it already](https://github.com/openjdk/jdk/pull/11940#discussion_r1095781185).

[Sergey is right](https://github.com/openjdk/jdk/pull/11940#discussion_r1099522374) there is *no contradiction*. It is how it is specified.

“The spec for `DONE` state does not specify when the transition occurs.” It says _after_, which is satisfied in either case: 1) `done` method is called, state transitions to `DONE`; or 2) state transitions to `DONE`, then `done` method is called.

The first case is how it was implemented before this fix. The second case is how it is implemented at this very moment in the this fix.


If you change the order of these two calls

https://github.com/openjdk/jdk/blob/76b02ea3a8add954705d6df7f91f5896d141b02d/src/java.desktop/share/classes/javax/swing/SwingWorker.java#L307-L308

you'll bring back the code to case 1 which aligns with the current specification.

If you do it, Sergey's concern [in this comment](https://github.com/openjdk/jdk/pull/11940#discussion_r1099522374) *will be resolved*. (For convenience: “Above we discussed that it is possible to see a difference if the listener will do some work on EDT, after the fix the listener will not be called last, is that safe to assume it does not break something?”)

Even though I don't see it as a problem, it is safer to preserve the specified order. Do you agree?

prsadhuk commented:
> I admit I got confused … by now.
> 
> It's been a long discussion.
> 
> > My take is current code fix and spec clarification is done to make code/spec wording consistent with **existing** spec whereas this change is to **enhance** the code/spec which I think we should keep it separate from this fix, as this will cause JCK test to be excluded and will cause JCK to update the test and I am not sure how that will pan out, so I suggest to keep the change to minimal to the extent only which is needed for this fix..
> 
> Yet the spec change is unnecessary if you change the order of calls as [discussed above](https://github.com/openjdk/jdk/pull/11940#discussion_r1095103200).
> 
> Referring to [the same thread](https://github.com/openjdk/jdk/pull/11940#discussion_r1095741184),
> 
> > > > It seems the order of sequence is listener->State.STARTED->doInBackground->listener->DONE->done
> > > 
> > > 
> > > Yes, but the different order is specified: listener(STARTED) -> doInBackground -> done -> listener(DONE).
> > 
> > 
> > But then it will violate
> > https://github.com/openjdk/jdk/blob/810c8a271b4524ae776e2306ef699e04a7d145a2/src/java.desktop/share/classes/javax/swing/SwingWorker.java#L288-L293
> 
> No, it doesn't violate: the state transitions to `DONE` _after method is finished_.
> 
> And I realised [I had addressed it already](https://github.com/openjdk/jdk/pull/11940#discussion_r1095781185).
> 
> [Sergey is right](https://github.com/openjdk/jdk/pull/11940#discussion_r1099522374) there is _no contradiction_. It is how it is specified.
> 
> “The spec for `DONE` state does not specify when the transition occurs.” It says _after_, which is satisfied in either case: 1) `done` method is called, state transitions to `DONE`; or 2) state transitions to `DONE`, then `done` method is called.
> 
> The first case is how it was implemented before this fix. The second case is how it is implemented at this very moment in the this fix.
> 
> If you change the order of these two calls
> 
> https://github.com/openjdk/jdk/blob/76b02ea3a8add954705d6df7f91f5896d141b02d/src/java.desktop/share/classes/javax/swing/SwingWorker.java#L307-L308
> 
> you'll bring back the code to case 1 which aligns with the current specification.
> 
> If you do it, Sergey's concern [in this comment](https://github.com/openjdk/jdk/pull/11940#discussion_r1099522374) _will be resolved_. (For convenience: “Above we discussed that it is possible to see a difference if the listener will do some work on EDT, after the fix the listener will not be called last, is that safe to assume it does not break something?”)
> 
> Even though I don't see it as a problem, it is safer to preserve the specified order. Do you agree?

Yes, I see..I have updated the fix and test.

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -64,14 +68,46 @@ protected String doInBackground() throws Exception {
+                               "PropertyChangeListeners called with " +
+                               "state STARTED before doInBackground is finished");
+                    }
+                    // After the doInBackground method is finished
```
Adding a blank line between the conditions would visually separate them.

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -64,14 +68,46 @@ protected String doInBackground() throws Exception {
-                if (!doInBackground) {
-                    throw new RuntimeException("done called before doInBackground");
+                if (!doInBackgroundFinished.get()) {
+                    throw new RuntimeException("done called before " +
+                                             " doInBackground is finished");
```
```suggestion
                    throw new RuntimeException("done called before " +
                                               "doInBackground is finished");
```

Add two spaces before the quote and to align it to the one on the line above; remove one space after the quote so that the error message doesn't have two consecutive spaces.

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -27,20 +27,23 @@
+    private static AtomicBoolean doInBackgroundFinished = new AtomicBoolean(false);
     private static final int WAIT_TIME = 200;
     private static final long CLEANUP_TIME = 1000;
     private static final CountDownLatch doneLatch = new CountDownLatch(1);
```
```suggestion
    private static final int WAIT_TIME = 200;
    private static final long CLEANUP_TIME = 1000;

    private static final AtomicBoolean doInBackgroundStarted = new AtomicBoolean(false);
    private static final AtomicBoolean doInBackgroundFinished = new AtomicBoolean(false);
    private static final CountDownLatch doneLatch = new CountDownLatch(1);
```

Separate the real constants from the objects which track the state. The flags should also be marked `final`.

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/SwingWorker.java:
```diff
@@ -304,13 +304,14 @@ public T call() throws Exception {
                 };
 
-       future = new FutureTask<T>(callable);
+        future = new FutureTask<T>(callable);
```
```suggestion
       future = new FutureTask<T>(callable);
```
An extra space breaks the correct indentation.

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -86,9 +122,10 @@ protected void done() {
+        System.out.println("doInBackground " + doInBackgroundFinished.get() +
                            " getState " + worker.getState());
-        if (doInBackground && worker.getState() != SwingWorker.StateValue.DONE) {
+        if (doInBackgroundFinished.get() &&
+              worker.getState() != SwingWorker.StateValue.DONE) {
```
```suggestion
        if (doInBackgroundFinished.get() &&
                worker.getState() != SwingWorker.StateValue.DONE) {
```
If you follow the style of double indentation on continuation lines, two more spaces should be added.

Consider moving the `&&` operator to the continuation line.

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -0,0 +1,134 @@
+                } catch (InterruptedException ex) {
+                    System.out.println("Got interrupted!");
+                }
+                try {
```
Maybe add a blank line between the two try-catch blocks?

And before `return`-statement?

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -0,0 +1,134 @@
+        if (doInBackgroundFinished.get() &&
+              worker.getState() != SwingWorker.StateValue.DONE) {
+            throw new RuntimeException("doInBackground is finished " +
+                                       " but State is not DONE");
```
```suggestion
                                       "but the State is not DONE");
```
Two consecutive spaces in an error message are redundant, one is enough.

The article seems missing?

prsadhuk replied:
Modified except condition kept separate for distinct failure message

aivanov-jdk reviewed (changes requested):
The changes look good except for the minor comments.

The updated fix doesn't require the CSR.

Please update the copyright year.

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -0,0 +1,134 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+/*
+ * @test
+ * @bug 8081474
+ * @summary  Verifies if SwingWorker calls 'done'
+ *           before the 'doInBackground' is finished
+ * @run main TestDoneBeforeDoInBackground
+ */
```
May I ask you to move the jtreg tags to the class declaration, please? When the file is opened in the IDE, they're not collapsed (together with the license block above) and easily accessible.

prsadhuk replied:
Not sure on this..it seems it is same for other tests also..

aivanov-jdk replied:
The majority of old tests have jtreg tags above imports. I find it very inconvenient.

Some newer tests have the jtreg tags before the class declaration.

jtreg accepts either variant.

When I review code, I always ask for this change because the jtreg tags are somewhat important and having them visible is just more convenient. The `@summary` tag provides the info similar to what one expects to find in a javadoc.

prsadhuk commented:
/csr unneeded

prsadhuk commented:
/csr unneeded

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -83,25 +86,27 @@ public void propertyChange(PropertyChangeEvent evt) {
                     // property change to StateValue.DONE
-                    if (doInBackgroundFinished.get() &&
-                            worker.getState() != SwingWorker.StateValue.DONE) {
+                    if (worker.getState() != SwingWorker.StateValue.DONE
+                          && doInBackgroundFinished.get()) {
```
```suggestion
                    if (worker.getState() != SwingWorker.StateValue.DONE
                        && doInBackgroundFinished.get()) {
```
This should also use four-space indentation on the continuation line for consistency with other cases.

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -83,25 +86,27 @@ public void propertyChange(PropertyChangeEvent evt) {
                     // property change to StateValue.STARTED
-                    if (doInBackgroundFinished.get() &&
-                         worker.getState() == SwingWorker.StateValue.STARTED) {
+                    if (doInBackgroundFinished.get()
+                        && worker.getState() == SwingWorker.StateValue.STARTED) {
```
```suggestion
                    if (doInBackgroundStarted.get()
                        && worker.getState() == SwingWorker.StateValue.STARTED) {
```
Align the condition to the error message.

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -83,25 +86,27 @@ public void propertyChange(PropertyChangeEvent evt) {
                     // before doInBackground started running
-                    if (doInBackgroundStarted.get() &&
-                         worker.getState() == SwingWorker.StateValue.STARTED) {
+                    if (doInBackgroundStarted.get()
+                        && worker.getState() == SwingWorker.StateValue.STARTED) {
```
```suggestion
                    if (doInBackgroundFinished.get()
                        && worker.getState() == SwingWorker.StateValue.STARTED) {
```
Align the condition to the error message.

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -124,10 +129,10 @@ public void propertyChange(PropertyChangeEvent evt) {
-              worker.getState() != SwingWorker.StateValue.DONE) {
-            throw new RuntimeException("doInBackground is finished " +
-                                       " but State is not DONE");
+        if (worker.getState() != SwingWorker.StateValue.DONE
+            && doInBackgroundFinished.get()) {
+              throw new RuntimeException("doInBackground is finished " +
+                                         "but State is not DONE");
```
```suggestion
        if (worker.getState() != SwingWorker.StateValue.DONE
            && doInBackgroundFinished.get()) {
            throw new RuntimeException("doInBackground is finished " +
                                       "but State is not DONE");
```
Throw shouldn't be indented by additional two spaces.

With the `&&` at the start of the line, it is rather clear where condition ends. Additionally, `throw` cannot be part of the condition.

It's a catch with `if` statement because its condition is always indented by four spaces. If required, a blank line could be added in the body of `if` to separate the condition visually; continuation line could use 8-space indentation.

prsadhuk commented:
I have tried alignment to the extent possible..I am not sure if it's required to be so finicky for it..

aivanov-jdk commented:
> I have tried alignment to the extent possible..

Not to the extent possible if it's different from the commonly used style.

> I am not sure if it's required to be so finicky about it :-)

It's probably not. However, the code which uses consistent style in the entire project is easier to read, it looks as if it was written by a single person even though it was actually written by a large team of different people. Otherwise, it's like a text with typos, you're still able to comprehend it but you stumble on each occurrence of one.

aivanov-jdk commented:
> > I have tried alignment to the extent possible..
> 
> Not to the extent possible if it's different from the commonly used style.
> 
> > I am not sure if it's required to be so finicky about it :-)
> 
> It's probably not.

In addition to that, in the cases I have provided suggestions to fix the indentation which can be applied with a single click.

prsadhuk commented:
I guess current iteration of the PR takes care of the alignment being suggested..

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -0,0 +1,139 @@
+        worker.addPropertyChangeListener(
+            new PropertyChangeListener() {
+                public void propertyChange(PropertyChangeEvent evt) {
+                    // Before doInBackground method is invoked,
+                    // SwingWorker notifies PropertyChangeListeners about the
+                    // property change to StateValue.STARTED
+                    if (doInBackgroundFinished.get()
+                        && worker.getState() == SwingWorker.StateValue.STARTED) {
+                        throw new RuntimeException(
+                               "PropertyChangeListeners called with " +
+                               "state STARTED after doInBackground is invoked");
+                    }
+
+                    // Ensure the STARTED state is notified
+                    // before doInBackground started running
+                    if (doInBackgroundStarted.get()
+                        && worker.getState() == SwingWorker.StateValue.STARTED) {
+                        throw new RuntimeException(
+                              "PropertyChangeListeners called with " +
+                               "state STARTED before doInBackground is finished");
+                    }
```
The comments, conditions and error messages do not match.

In the first case, it should be `doInBackgroundStarted` instead of `doInBackgroundFinished`; in the second — `doInBackgroundFinished` correspondingly.

prsadhuk replied:
Test is updated to test the order

aivanov-jdk reviewed (changes requested):
Should `evt.getNewValue()` be used to identify the new value of the state property?

The test does not fail if change the order of the lines in `finally` block that changes the order in which the `done` method is called and listeners are notified. I expected it to fail. Isn't it the reason why `PropertyChangeListener` was added?

aivanov-jdk commented on test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java:
```diff
@@ -0,0 +1,147 @@
+        worker.addPropertyChangeListener(
+            new PropertyChangeListener() {
+                public void propertyChange(PropertyChangeEvent evt) {
+                    System.out.println("doInBackground started: " +
+                                        doInBackgroundStarted.get() +
```
```suggestion
                    System.out.println("doInBackgroundStarted: " +
                                        doInBackgroundStarted.get() +
```
Just for consistency with the following `-Finished` line.

aivanov-jdk reviewed (approved):
I would appreciate if you update the copyright year in `SwingWorker`.

mrserb commented on src/java.desktop/share/classes/javax/swing/SwingWorker.java:
```diff
@@ -301,17 +301,16 @@ public SwingWorker() {
+                            return doInBackground();
+                        } finally {
+                            doneEDT();
+                            setState(StateValue.DONE);
```
Can the test checks the case when the listener called from "setState" will throw an exception, I think some of the previous iterations of the fix will not work in this case.

aivanov-jdk replied:
> Can the test checks the case when the listener called from "setState" will throw an exception, I think some of the previous iterations of the fix will not work in this case.

This is not a new problem and _is not introduced_ by Prasanta's fix.

I also found it while I was testing his changes and ran the test.

If a listener throws an exception, all the following runnables which are already accumulated in `AccumulativeRunnable` are not run.

I have already written a test for this case and I will submit a bug. It must not be included here.

mrserb replied:
I think the previous iteration of the fix was not run the doneEDT if the setState throws an exception, while the current version, and initial code always run it.

aivanov-jdk replied:
@mrserb I have submitted [JDK-8302778](https://bugs.openjdk.org/browse/JDK-8302778): _SwingWorker loses notifications if a handler throws an exception_.

The provided [LostNotifications.java](https://github.com/aivanov-jdk/jdk/blob/e4042edc654849a992710a0e0d23ec8df4fe14c4/test/jdk/javax/swing/SwingWorker/LostNotifications.java) test fails on 8, 11, 17 and 21.

It proves that this problem is not related to the current fix, however, the test for this issue has uncovered it.

mrserb replied:
That look fine.

prsadhuk commented:
/integrate
